modules/apps/remote-app/remote-app-client-js/src/main/resources/META-INF/resources/js/index.ts

```diff
@@ -0,0 +1,897 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {default as logger} from './log';
+
+/**
+ * Subset of the full-fledged `Response` type that can actually be serialized
+ * across iframes.
+ */
+type BasicResponse = Pick<
+	Response,
+	| 'blob'
+	| 'json'
+	| 'ok'
+	| 'redirected'
+	| 'status'
+	| 'statusText'
+	| 'text'
+	| 'url'
+> & {
+	headers: Pick<Headers, 'get' | 'has'>;
+};
+
+type ClientOptions = {
+	debug: boolean;
+};
+
+/**
+ * Payloads may only contain properties which we can send across iframes
+ * using the `postMessage` API. Complex properties like functions cannot
+ * be sent because the "Structured clone algorithm" does not support
+ * them.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Client/postMessage
+ */
+type Payload = {[key: string]: StructuredClonable};
+
+type PromiseMap<T = unknown> = {
+	[key: string]: {
+		reject: (reason?: any) => void;
+		resolve: (value?: T) => void;
+	};
+};
+
+/**
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
+ */
+type StructuredClonable =
+	| Array<StructuredClonable>
+	| ArrayBuffer
+	| ArrayBufferView
+	| Blob
+	| Boolean
+	| Date
+	| File
+	| FileList
+	| ImageBitmap
+	| ImageData
+	| Map<StructuredClonable, StructuredClonable>
+	| RegExp
+	| Set<StructuredClonable>
+	| String
+	| boolean
+	| null
+	| number
+	| string
+	| undefined
+	| {[key: string]: StructuredClonable};
+
+type State =
+	| 'disposed'
+	| 'invalid'
+	| 'registered'
+	| 'registering'
+	| 'unregistered';
+
+type Subscription = {
+	release: () => void;
+};
+
+// TODO consider moving these into a shared location
```

Comment:
You'll see quite a few TODO comments scattered around here, because draft and all that.